streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregateTest.java

```diff
@@ -0,0 +1,466 @@
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(6, 16)), "0+3+4+1", 16),
+                        // C@11 right window updated when C@21 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(12, 22)), "0+4+1+1", 21),
+                        // C@15 right window updated when C@21 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(16, 26)), "0+1+1", 21),
+                        // C@16 right window created when C@21 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(17, 27)), "0+1", 21),
+                        // C@21 left window created when C@21 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(11, 21)), "0+3+4+1+1", 21),
+                        // C@15 right window updated when C@23 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(16, 26)), "0+1+1+1", 23),
+                        // C@16 right window updated when C@23 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(17, 27)), "0+1+1", 23),
+                        // C@21 right window created when C@23 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(22, 32)), "0+1", 23),
+                        // C@23 left window created when C@23 processed
+                        new KeyValueTimestamp<>(new Windowed<>("C", new TimeWindow(13, 23)), "0+4+1+1+1", 23),
+
+                        // D@11 left window created when D@11 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(1, 11)), "0+4", 11),
+                        // D@11 right window created when D@12 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(12, 22)), "0+2", 12),
+                        // D@12 left window created when D@12 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(2, 12)), "0+4+2", 12),
+                        // D@29 left window created when D@29 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(19, 29)), "0+3", 29),
+                        // D@11 right window updated when D@16 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(12, 22)), "0+2+5", 16),
+                        // D@12 right window created when D@16 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(13, 23)), "0+5", 16),
+                        // D@16 left window created when D@16 processed
+                        new KeyValueTimestamp<>(new Windowed<>("D", new TimeWindow(6, 16)), "0+4+2+5", 16)
+                        ),
+                supplier.theCapturedProcessor().processed
+        );
+    }
+
+    @Test
+    public void testJoin() {
```

Comment:
I still don't exactly understand why we have a join test in the `KStreamXXWindowAggregateTest`, but thanks for adding it for sliding windows. I'm sure there was a good reason for it, probably long ago

Reply:
The Achilles heel of implementing new KTable features has historically been that we forgot to test them in a context that required the ValueGetter to work properly, of which Join is a notable use case. I'd actually say it should be required for every KTable operator to have a test where it's the source of a Join. For stateless operators, we should test both with and without a Materialized argument on the operator.

Reply:
Aha, so there was a good reason for it 😄